lib/Serialization/DeserializeSIL.cpp

```diff
@@ -410,11 +412,20 @@ SILFunction *SILDeserializer::getFuncForReference(StringRef name,
     }
   }
 
-  // FIXME: check for matching types.
-
   // At this point, if fn is set, we know that we have a good function to use.
-  if (fn)
+  if (fn) {
+    SILType fnType = fn->getLoweredTypeInContext(context);
+    if (fnType != type) {
+      StringRef fnName = fn->getName();
+      if (auto *dc = fn->getDeclContext()) {
+        if (auto *decl = dyn_cast_or_null<AbstractFunctionDecl>(dc->getAsDecl()))
+          fnName = decl->getNameStr();
+      }
+      fn->getModule().getASTContext().Diags.diagnose(fn->getLocation().getSourceLoc(), diag::deserialize_function_type_mismatch, fnName, fnType.getASTType(), type.getASTType());
+      exit(1);
```

Comment:
Is there something we can do other than `exit`?

Reply:
I didn't find anything better. We need to abort here immediately to not run into subsequent asserts/crashes.
I'm open to suggestions :-)

Reply:
Maybe llvm::report_fatal_error()?

Reply:
That also let it look like a compiler crash